### PR TITLE
Multi-Tensor splitting and stacking utilities

### PR DESCRIPTION
**Pull request by mrossinek**
### Summary

In a recent project I had to implement a lot of array stacking and splitting.
This use case cannot be handle by our `.apply` utilities of `PolynomialTensor` and `ElectronicIntegrals` because:
- in the stacking case the numpy routines expect the arrays to be stacked provided as a list rather than separate positional arguments
- in the splitting case we return multiple tensors instead of one

The latter case would have been possible to implement in a "multi-apply" setting (which I plan to add in the future, too) but splitting arrays is fairly common so I think it deserves its own utility.

### CODE REVIEW

**Comment by mrossinek** [comment]
I rebased this on top of #1214 for the time being. Once that is merged (and CI fixed accordingly except for Neko) I will rebase back on `main` :+1:

File: qiskit_nature/second_q/operators/electronic_integrals.py
Comment (ftroisi) [inline qiskit_nature/second_q/operators/electronic_integrals.py:529]: According to the definitions of `alpha`, `beta` and `beta_alpha` I imagine that the former two typically have the same length, but that's not the case the `beta_alpha`. If so, why are the same `indices_or_sections` used to split all three PolynomialTensors? Is it possible to have a case where we ask to spilt the Tensor in an out-of-bound index?

File: qiskit_nature/second_q/operators/electronic_integrals.py
Comment (ftroisi) [inline qiskit_nature/second_q/operators/electronic_integrals.py:504]: Perhaps some unit tests should be added for this method. I know that this method heavily relies on the same one in the PolynomicTensor (which is tested), but still it would be good to test it. Especially due to this special behavior

File: qiskit_nature/second_q/operators/electronic_integrals.py
Comment (ftroisi) [inline qiskit_nature/second_q/operators/electronic_integrals.py:455]: See comment about unit tests in split method

File: qiskit_nature/second_q/operators/electronic_integrals.py
Comment (ftroisi) [inline qiskit_nature/second_q/operators/electronic_integrals.py:463]: Please add the warning present in PolynomialTensors: `Disable this setting with care!`

File: qiskit_nature/second_q/operators/electronic_integrals.py
Comment (ftroisi) [inline qiskit_nature/second_q/operators/electronic_integrals.py:512]: Please add the warning present in PolynomialTensors: `Disable this setting with care!`

File: qiskit_nature/second_q/operators/electronic_integrals.py
Comment (ftroisi) [inline qiskit_nature/second_q/operators/electronic_integrals.py:508]: Perhaps we should explicit about the two numpy function that will be most commonly used (`hsplit`, `vsplit`).
This applies also to the stack method

**Review by ftroisi — COMMENTED** [review]
Overall it looks good to me, but I'm not an expert on PolynomialTensors so a second pair of eyes would be good

File: qiskit_nature/second_q/operators/electronic_integrals.py
Comment (mrossinek) [reply qiskit_nature/second_q/operators/electronic_integrals.py:529]: Why do you say that the lengths will not be the same? All these matrices will have an "edge length" of the number of molecular orbitals. The `alpha` and `beta` PolynomialTensors will usually have 1-body and 2-body terms (typically stored in `+-` and `++--`) while `beta_alpha` only has a 2-body matrix.
However, (at least how I use these methods) I am mostly interested in the manipulation of MO coefficients which only have 1-body terms in `alpha` and `beta` and no mixing terms present.

Nonetheless, the length should always match.

It will be up to the user to ensure they split on valid indices there is nothing we can do in this regard here.

File: qiskit_nature/second_q/operators/electronic_integrals.py
Comment (mrossinek) [reply qiskit_nature/second_q/operators/electronic_integrals.py:504]: Yeh I was kind of in a hurry. I will see if I can add some unittests similar to how it is done in the `einsum` routine.

File: qiskit_nature/second_q/operators/polynomial_tensor.py
Comment (woodsp-ibm) [inline qiskit_nature/second_q/operators/polynomial_tensor.py:652]: How much is `validate` actually valid to use in these scenarios. For instance this will fail if validate is True. The only way I can see it passing validate is if you stacked here 2 arrays that individually themselves would fail. I note all the examples  here set it false, as do the unittests. Should we rather default it to false on the stack method or remove the parameter all together so internally tensors are created with validation off?

Comment (woodsp-ibm) [reply qiskit_nature/second_q/operators/polynomial_tensor.py:652]: Fair enough - but then would it be better to say something a little different about `validate` in stack and split, than in the constructor? Maybe for split one is more likely going from something that would not validate to something that will. But stating something like the sentiment you expressed above in the stack method, maybe even as a Note to highlight things could be helpful.  By default its likely to fail as per that example - we may intend for this to be the case to ensure a user really knows what they are doing, but some text to that effect may help. For split I guess it usually ought to pass right.

File: qiskit_nature/second_q/operators/electronic_integrals.py
Comment (ftroisi) [reply qiskit_nature/second_q/operators/electronic_integrals.py:532]: Ok I think I got the point. Should be alright then!

File: qiskit_nature/second_q/operators/polynomial_tensor.py
Comment (mrossinek) [reply qiskit_nature/second_q/operators/polynomial_tensor.py:652]: Splitting is just as likely to lead to mismatching shapes. I will add a note to both of the docstrings.

File: qiskit_nature/second_q/operators/electronic_integrals.py
Comment (woodsp-ibm) [reply qiskit_nature/second_q/operators/electronic_integrals.py:521]: In the output from the documentation built here these unfortunately did not end up as links - just highlighted - since the url for say hsplit is https://numpy.org/doc/stable/reference/generated/numpy.hsplit.html I wonder whether here if its says `numpy.hsplit` whether that resolves fine

File: qiskit_nature/second_q/operators/electronic_integrals.py
Comment (woodsp-ibm) [inline qiskit_nature/second_q/operators/electronic_integrals.py:472]: Still not working - I think it was my bad, sorry - I said `meth` but for these they should be `func` I tried locally in PolyTensor just adding a couple of func refs and it comes out fine  so should here - I did try with np as well but it does not seem to resolve that in this context so ``:func:`numpy.hstack` `` etc and it should all be good.

**Review by woodsp-ibm — APPROVED** [review]

**Pull request merged**
